Performance: MetricEventManager _runningFunctions to ConcurrentDictionary

NickCraver opened the pull request:
Part of #7908 - `_runningFunctions` currently causes locking under load especially when a metrics flush happens to cull the list. I caught up on history with @mathewc and @brettsam here. It seems a combination of `ConcurrentDictionary` semantics and previous circular refs were the reasoning for going this way in the past. The last change to this was in #6547, but since then `ConcurrentDictionary` has improved and we can further improve the mechanism by not using a string allocation on the key front. Since an `InvocationId` is already unique we can key off it and opportunistically remove it.

Overall: this moves back to the original storage structure with a tuned key approach now that .NET has changed the locking bits internally (since 5.x - see https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-5/#collections) and it's no longer a hot path in profiling.

Note that I kept the culling for safety as the `List<T>` approach did - which ultimately may have been one cause of the previous memory leaks - an incomplete function could manifest as a hung reference. This may be impossible now allowing us to remove it, but it's not causing harm and is of minimal cost to keep the security here rather than relying on perfection in the pipe.

### Issue describing the changes in this PR

Resolves a component of #7908

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

NickCraver committed:
Performance: MetricEventManager _runningFunctions to ConcurrentDictionary

Part of #7908 - `_runningFunctions` currently causes locking under load especially when a metrics flush happens to cull the list. I caught up on history with @mathewc and @brettsam here. It seems a combination of `ConcurrentDictionary` semantics and previous circular refs were the reasoning for going this way in the past. The last change to this was in #6547, but since then `ConcurrentDictionary` has improved and we can further improve the mechanism by not using a string allocation on the key front. Since an `InvcocationId` is already unique we can key off it and opportunistically remove it.

Overall: this moves back to the original storage structure with a tuned key approach now that .NET has changed the locking bits internally (since 5.x - see https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-5/#collections) and it's no longer a hot path in profiling.

brettsam commented:
I went back and looked -- I was wrong about why we moved away from `ConcurentDictionary` here. What I was remembering was some issues we saw way back with `ConcurrentBag` (see here -- https://github.com/Azure/azure-webjobs-sdk/pull/1971#issue-371703839).

brettsam commented on src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs:
```diff
@@ -462,30 +460,32 @@ internal void Flush()
-                    runningFunctionsSnapshot = _runningFunctions.ToArray();
+                    if (possiblyRunning.Value.Completed)
+                    {
+                        _runningFunctions.TryRemove(possiblyRunning.Key, out _);
```
Does this not throw a `Collection was modified...` exception?

NickCraver replied:
Nope! We're safe since it's a concurrent collection now. This safety mechanism is even _likely_ to race as the metrics flush will often overlap a function's own end statement but that race should be safe and not cause any issues. Please sanity check my thinking there!

brettsam replied:
Yup -- you're right! https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getenumerator?view=net-6.0#remarks

danielmarbach commented on src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs:
```diff
@@ -462,30 +460,32 @@ internal void Flush()
 
-                // we calculate concurrency here based on count, since these events are raised
+                // If not for the length, we could probably avoid this array allocation
+                var concurrency = _runningFunctions.Count;
```
Doesn't count acquire all the locks? Since you have already iterated through the collection above why not simply count there and than use that as the concurrency value or does the value need to be "precise" at the moment when you take it possibly including things that have been added in the meantime? I guess since there is no atomicity between those two iterations anyway and you take a snapshot of count counting above would be sufficient

danielmarbach commented on src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs:
```diff
@@ -462,30 +460,32 @@ internal void Flush()
                 // the current count.
                 var currentTime = DateTime.UtcNow;
-                foreach (var runningFunction in runningFunctionsSnapshot)
+                foreach (var runningFunction in _runningFunctions.Values)
```
Values also acquires the locks and creates a new collection/snapshot. Is that necessary, or could another enumeration be good enough?

NickCraver replied:
We could get more complicated here, but we'd also quickly get into diminishing returns vs. maintenance of something more complex, IMO. While `GetValues()` under the hood does a lock acquisition and snapshot, we're generally talking about very few items here. The cost amortization is also important - this runs once every `n` seconds to dump based on config but with the new cleanup only contains the running values (+/- instantaneous delta inside any method), then we're only locking _once per cycle_ rather than a lock on every op (where we see a great deal of contention in profiling).

Talking through this though - we _could_ change it up. We don't need _another_ enumeration though, we've already got one above. If we're creating the same list of items _anyway_ we could just pretend we're microseconds earlier and go with an `else` in the loop above, creating our "running" collection in the existing enumeration. It'd eat the same number of resizes in the list (still relatively rarely in the overall scheme) as the current `.Values` call.

NickCraver replied:
@danielmarbach updated :)

NickCraver committed:
Remove .Values lock

These no measurable difference today in throughput, but this is a tad bit simpler to reason about and eliminates another iteration, so let's do it. Good call @danielmarbach :)

danielmarbach commented on src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs:
```diff
@@ -462,30 +460,37 @@ internal void Flush()
-                lock (_runningFunctionsSyncLock)
+                // Loop through and prune any completed runs
+                // Also collects the currently running functions in the same enumeration to minimize cost.
+                var running = new List<FunctionStartedEvent>();
```
Depending how hot this path is the list could be Pooled. Hard to say though without more context of that is worth it

mathewc commented on src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs:
```diff
@@ -333,15 +333,14 @@ private class FunctionActivityTracker : IDisposable
             private int _functionActivityFlushInterval;
             private ConcurrentQueue<FunctionMetrics> _functionMetricsQueue = new ConcurrentQueue<FunctionMetrics>();
-            private List<FunctionStartedEvent> _runningFunctions = new List<FunctionStartedEvent>();
+            private ConcurrentDictionary<Guid, FunctionStartedEvent> _runningFunctions = new ConcurrentDictionary<Guid, FunctionStartedEvent>();
```
We had intentionally moved away from a ConcurrentDictionary due to hot lock issues in performance investigations. The changes we made resolved that. My concern is that we may regress here. The relevant PR was https://github.com/Azure/azure-functions-host/pull/6547 where we had moved from ConcurrentDictionary to a List

NickCraver replied:
Yep agreed - I dug into this more as well and tried to add detail in the PR description. The locking internals have changed since when those changes have made - given we're on .NET 5+ this is a much less contentious path and results in more throughput.

I'd love to get the anon/benchmark bits in so we can add Crank data to each of the PRs here - thoughts on https://github.com/Azure/azure-functions-host/pull/7941#issuecomment-986947273? It'd make it much easier to provide numbers for each, though they're all small and add-up I totally agree it's worthwhile.

NickCraver committed:
Merge branch 'dev' into craver/metrics-running-dictionary

NickCraver commented:
Superseded by #7936 (requires both to really test and benchmark the changes since we have 2 overlapping locks here).
